x-pack/plugins/canvas/public/lib/aeroelastic/types.ts

```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+type f64 = number; // eventual AssemblyScript compatibility; doesn't hurt with vanilla TS either
+type f = f64; // shorthand
+
+export type vector2d = [f, f, f] & ReadonlyArray<f> & { __nominal: 'vector2d' };
+export type vector3d = [f, f, f, f] & ReadonlyArray<f> & { __nominal: 'vector3d' };
+
+export type transformMatrix2d = [f, f, f, f, f, f, f, f, f] &
+  ReadonlyArray<f> & { __nominal: 'transformMatrix2d' };
+export type transformMatrix3d = [f, f, f, f, f, f, f, f, f, f, f, f, f, f, f, f] &
```

Comment:
@clintandrewhall I internally named it the solution [in Spanish](https://www.youtube.com/watch?v=Yx4ABAKyo-A) - I tried a bunch of alternatives with fixed-length arrays but haven't yet succeeded. Here's a [playground](https://tinyurl.com/yd6q4zff) which shows what ought to pass and fail. I tried to abstract out the array length eg. `&`-ing or `extend`ing with 
```
{
  0: T;
  length: L;
}
```
but no approach I tried worked out. Also, we have only these 4 array lengths so maybe it's OK. 

Why I bothered with nominal typing? In the future, we may have row-major matrices in addition to the current column-major matrices, their length and element type (ie. everything 🦆) would be identical. Yet, multiplying a row major and a column major matrix doesn't make sense with a function that expects eg. two column major matrices. 

Also, it's good to distinguish between the `vector3d` here (a point in an affine 3d space, ie. 4 elements) from stuff we have in `layout_functions.js` for representing x/y bounds, possibly also a 4-element numerical array.

IOW aiming for duck typing alone would feel like reaching 30% of the objectives with 90% of the work 😄